ls1intum/Artemis
# `Development`: Refactor feedback generation for programming exercises

### Checklist
#### General


- [x] **All** changes and their related features were tested with **all** corresponding user types on a test server by other developers.
#### Server

### Motivation and Context

The `JenkinsService` and `BambooService` do nearly the same in the `addFeedbackToResult()` method. This was already marked with a note as a future todo.

As part of adding support for an additional CI system, duplicating this method for a third time was no solution we would be happy with.

### Description
We have moved the `addFeedbackToResult()` method to the common `AbstractContinuousIntegrationService` super class. This contains all the relevant functionality. 

The Continuous Integration System specific DTOs implement newly created interfaces, which define methods needed by `addFeedbackToResult()` (e.g. as the list of successful test cases and extracting the message of a test case).

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Students
- 1 Existing Programming Exercise

1. With the student, participate in a programming exericse, make code changes (make sure a test case will fail) and push them. In Artemis feedback should be displayed. Fix the issue so that the tests will pass and push the changes again.
2. Do the same with a newly created programming exercise with static code analysis and testwise coverage enabled. Make sure, the static code analysis reports and testwise coverage reports are processed.

### Review Progress

#### Code Review
#### Manual Tests

### Test Coverage
| Class/File | Branch | Line |
|------------|-------:|-----:|
| `AbstractContinuousIntegrationService.java` | 88% | 100% |
| `BambooService.java` | 56% | 80% |
| `BambooBuildResultNotificationDTO.java` | 100% | 100% |
| `JenkinsService.java` | 83% | 98% |
| `TestCaseDTO.java` | 85% | 81% |
| `TestResultsDTO.java` | 75% | 98% |
| `TestSuiteDTO.java` | 100% | 100% |
| `AbstractBuildResultNotificationDTO.java` | 75% | 91% |
